src/Application.php

```diff
@@ -213,11 +213,11 @@ public function routeMiddleware(ServerRequestInterface $request, ResponseInterfa
      * @param callable|string $middleware Middleware (or middleware service name) to associate with route.
      * @param null|array $methods HTTP method to accept; null indicates any.
      * @return Router\Route
-     * @throws InvalidArgumentException if $path is not a Router\Route AND middleware is null.
+     * @throws InvalidArgumentException if $path is not a Router\Route|string AND middleware is null.
      */
     public function route($path, $middleware = null, array $methods = null)
     {
-        if (! $path instanceof Router\Route && null === $middleware) {
+        if ((! is_string($path) || ! $path instanceof Router\Route) && null === $middleware) {
             throw new InvalidArgumentException(sprintf(
```

Comment:
This refactor is incorrect. These are the possible signatures:

- `route(Router\Route $route)`
- `route(string $route, $middleware, array $methods = null)`

This is why the conditional looked for the combination of "not a Route instance AND middleware is missing".

As for the exception when `$path` is not a string, that is raised by the `Router\Route` instance itself when we attempt to instantiate one at [line 279](https://github.com/zendframework/zend-expressive/blob/master/src/Application.php#L279) (see [lines 63-65 of the Route class](https://github.com/zendframework/zend-expressive/blob/master/src/Router/Route.php#L63-L65)).

I also already have tests in place for these behaviors.